src/promise.js

```diff
@@ -44,3 +44,38 @@ export function all(promises) {
     });
   });
 }
+
+
+/**
+ * Returns a promise resolver - a construct that contains the promise itself
+ * as well as resolve and reject functions.
+ * @return {!Resolver<T>}
+ * @template T
+ */
+export function newResolver() {
+  let resolve, reject;
+  const promise = new Promise((aResolve, aReject) => {
+    resolve = aResolve;
+    reject = aReject;
+  });
+  return new Resolver(promise, resolve, reject);
+}
+
+/**
+ * @template T
+ */
+class Resolver {
```

Comment:
can we rename this to `Deferred` as its close to the jquery Deferred object. (where its an instance that has a .promise property and .resolve/.reject methods). i dont feel strongly either way so feel free to ignore.